### PR TITLE
Rework how the remappings middleware matches remapped to/from columns for explicit external (FK) remaps

### LINKED ISSUE #9236: Display value is not honored when there are more than one foreign key to the same table
**Problem:**

When a table has multiple foreign key attributes to the same table, the Display Values of these attributes are not correctly honorred. It appears that only one of them is correctly handled.

**Expected behaviour:**

I would expect that both columns display values to show the foreign key name.

**Steps to reproduce:**

- Create the following tables and data:
```
CREATE TABLE Products (
    id int,
    name varchar(255),
    family int,
    sub_family int,
    PRIMARY KEY (id)
);

CREATE TABLE Family (
    id int,
    name varchar(255),
    PRIMARY KEY (id)
);

INSERT INTO Family VALUES (1,'Vegetable');
INSERT INTO Family VALUES (2,'Orange');
INSERT INTO Family VALUES (3,'Banana');
INSERT INTO Family VALUES (4,'Meat');
INSERT INTO Family VALUES (5,'Pork');
INSERT INTO Family VALUES (6,'Chicken');

INSERT INTO Products VALUES (1,'Orange Juice', 1, 2);
INSERT INTO Products VALUES (2,'Banana pie', 1, 3);
INSERT INTO Products VALUES (3,'Chicken nuggets', 4, 6);
```

- Then configure the Products table in Metabase so that it looks like this:
image

- Configure both Family and SubFamily attributes so that the Display Value is obtained through the foreign key:
image

- Now perform a question on the Products table. The output is as follow:
image

For some reason the Family attribute is showing the ID instead of showing the foreign key name. This only occurs if the foreign key table for both attributes is the same. If I disable one of the column or the other, it will correctly show the display value. However, when both columns are visible in the query, only one of them is correctly showing the values:
image
image

**Additional info:**
- Your browser and the version: Chrome  71.0.3578.98
- Your operating system: Windows 10
- Your databases: Postgres
- Metabase version: 0.31.2
- Metabase hosting environment: Linux
- Metabase internal database: Postgres

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by camsaul**
Fixes #9236

_Note to reviewer: if you're unfamiliar with remaps ("dimensions") I suggest reading the docstring for [`metabase.models.params.chain-filter`](https://github.com/metabase/metabase/blob/master/src/metabase/models/params/chain_filter.clj#L1) first where I wrote down all of this stuff in great detail a while back. This PR only concerns explicit external (FK) remaps. Internal remaps' behavior is not changed._

Previously the QP remapping middleware looked at Field `:id` to determine the remapped-from and remapped-to columns which failed when the same column was remapped to more than once (e.g. #9236). After trying a bunch of things to make this work I finally settled on just recording which Fields we remap to/from by adding new `::original-field-dimension-id` and `::new-field-dimension-id` keys to the `:field` clause options to mark which the `Dimension` `:id` for the original Field that is getting remapped and newly added remapped-to column, respectively. To make this available in post-processing I tweaked the `annotate` middleware to add a new key to metadata, `:options`, that includes any namespaced `:field` clause options at the end of preprocessing, and to remove namespaced options from the `:field_ref`.

Other that all the tests that need to be updated the fix is working as expected.

### CODE REVIEW

**Comment by camsaul** [comment]
Please review 😭

File: shared/src/metabase/mbql/util.cljc [hunk 1]
```diff
@@ -680,6 +680,13 @@
   (assert (is-clause? :field clause))
   (assoc-field-options clause :temporal-unit unit))
 
+(defn remove-namespaced-options
+  "Update a `:field`, `:expression` reference, or `:aggregation` reference clause by removing all the "
```
Comment (howonlee) [inline shared/src/metabase/mbql/util.cljc:684]: unfinished sentence?

File: test/metabase/test_runner/effects.clj [hunk 2]
```diff
@@ -45,10 +45,22 @@
        :diffs    (when-not pass?#
                    [[actual# [(s/check schema# actual#) nil]]])})))
 
+(defn derecordize
```
Comment (howonlee) [inline test/metabase/test_runner/effects.clj:48]: wait, where's the side effect? this one looks pure?

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 3]
```diff
@@ -23,112 +23,189 @@
...
+  [{{:keys [source-query]} :query, :as query}]
+  (if (and source-query (not (:native source-query)))
+    ;; Only do lifting if source is MBQL query
+    (let [{source-query-remaps :remaps, source-query :query} (add-fk-remaps (assoc query :query source-query))]
+      {:remaps source-query-remaps
+       :query  (assoc-in query [:query :source-query] (:query source-query))})
+    {:remaps nil, :query query}))
+
+(s/defn ^:private add-fk-remaps :- QueryAndRemaps
+  "Add any Fields needed for `:external` remappings to the `:fields` clau se of the query, and update `:order-by`
```
Comment (howonlee) [inline src/metabase/query_processor/middleware/add_dimension_projections.clj:181]: typo

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 4]
```diff
@@ -280,21 +431,23 @@
      ;; Get the entries we're going to add to `:cols` for each of the remapped values we add
      :internal-only-cols (map :new-column internal-only-dims)}))
 
-(defn- add-remapped-cols
+(s/defn ^:private add-remapped-cols
```
Comment (howonlee) [inline src/metabase/query_processor/middleware/add_dimension_projections.clj:434]: it annoys me that `add-remapped-cols` and `add-remapped-columns` has basically different semantics and are for basically different parts of the remapping process.

File: src/metabase/query_processor/middleware/add_dimension_projections.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/query_processor/middleware/add_dimension_projections.clj]: fixed

File: src/metabase/query_processor/middleware/add_dimension_projections.clj (hunk 4 above)
Comment (camsaul) [reply src/metabase/query_processor/middleware/add_dimension_projections.clj]: that's a good point. I renamed it to make it less confusing

File: test/metabase/test_runner/effects.clj (hunk 2 above)
Comment (camsaul) [reply test/metabase/test_runner/effects.clj:35]: You're right. This namespace doesn't actually have any side effects anymore. It did at one point but all that stuff was moved out a while ago. I went ahead and renamed the namespace to reflect that.

File: shared/src/metabase/mbql/util.cljc (hunk 1 above)
Comment (camsaul) [reply shared/src/metabase/mbql/util.cljc]: Fixed

**Review by howonlee — APPROVED** [review]
It works fine, tests work, style is plain enough that I can read it.

Didn't Dan worry about the perf of all these matches we're doing now? Maybe we should have profiling or something shoved in our sdlc somewhere

**Comment by camsaul** [comment]
Yeah, @dpsutton mentioned that the other day. My take is that the `core.match` stuff compiles down to reasonably well optimized code and even if we walk and rewrite the query a hundred times while compiling the query it's going to be pretty small compared to the overhead from actually running the query and transducing the result rows. But we can definitely add some profiling around that somewhere and then maybe write more optimized search-and-replace methods in the future

**Pull request merged**